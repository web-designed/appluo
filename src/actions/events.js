import uuid from 'uuid'
import moment from 'moment'
import database from '../firebase/firebase'


//*******************************************************
// EVENTS ACTION GENERATORS
//*******************************************************

export const addEvent = (event) => ({
   type:'ADD_EVENT',
   event
})

export const startAddEvent = (eventData = {}) => {
   return (dispatch) => {
      const {
         note = '', 
         cleaner = 'unknown', 
         cleanedAt = 0,
         createdAt = 0,
         place = ''
      } = eventData

      const event = { note, cleaner, cleanedAt, createdAt, place }

      return database.ref('events').push(event).then((ref) => { // add return to chain promises in the tests
         dispatch(addEvent({
            id: ref.key,
            ...event
         }))
      })
   }
}


export const removeEvent = (id) => ({
   type: 'REMOVE_EVENT',
   id
})

export const editEvent = (id, update) => ({
   type: 'EDIT_EVENT', 
   id,
   update
})

export const addComment = (eventId, { commentBody, commenter }) => ({
   type: 'ADD_COMMENT',
   eventId,
   comment: {
      id: uuid(),
      createdAt: moment().valueOf(),
      commenter,
      commentBody
   }
})

export const removeComment = (eventId, commentId) => ({
   type:'REMOVE_COMMENT',
   eventId,
   commentId
})

export const editComment = (eventId, commentId, update) => ({
   type:'EDIT_COMMENT',
   eventId,
   commentId,
   update
})