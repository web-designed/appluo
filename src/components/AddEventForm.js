import React from 'react'
import { connect } from 'react-redux'
import uuid from 'uuid'
import moment from 'moment'
import 'react-dates/initialize'
import { SingleDatePicker } from 'react-dates'
import { Link } from 'react-router-dom'

export class AddEventForm extends React.Component {

   state = {
      date: this.props.cleanedAt ? moment(this.props.cleanedAt) : moment(),
      focused: false,
      note: this.props.note ? this.props.note : '',
      place: this.props.place ? this.props.place : this.props.settings.places[0],
      error: '',
      cleaner: this.props.cleaner ? this.props.cleaner : this.props.settings.cleaners[0],
   }

   onCleanerChange = (e) => {
      const cleaner = e.target.value
      this.setState(() => ({ cleaner }))
   }

   onPlaceChange = (e) => {
      const place = e.target.value
      this.setState(() => ({ place }))
   }

   onNoteChange = (e) => {
      const note = e.target.value
      this.setState(() => ({ note }))
   }

   onDateChange = (date) => {
      this.setState(() => ({ date }))
   }

   onFocusChange = ({focused}) => {
      this.setState(() => ({focused}))
   }

   handleSubmit = (e) => {

      e.preventDefault()
      
      const event = {
         cleanedAt: this.state.date.valueOf(),
         createdAt: moment().valueOf(),
         cleaner: this.state.cleaner,
         place: this.state.place,
         note: this.state.note,
      }

      this.props.handleSubmit(event)
   }

   render(){

      console.log(this.props)

      return(
         <div>
            <form onSubmit={this.handleSubmit}>
               <div class="input-group mb-2">
                  <div class="input-group-prepend">
                     <label class="input-group-text" for="cleaner">Name</label>
                  </div>
                  <select 
                     class="custom-select" 
                     id="cleaner"
                     name="cleaner" 
                     onChange={this.onCleanerChange}
                     value={this.state.cleaner.toLowerCase()}
                  >
                     {
                        this.props.settings.cleaners.map( cleaner => {
                           return <option value={cleaner.toLowerCase()} key={cleaner.toLowerCase()}>{cleaner}</option>
                        })
                     }
                  </select>
               </div>
               <div class="input-group mb-2">
                  <div class="input-group-prepend">
                     <label class="input-group-text" for="place">Place</label>
                  </div>
                  <select 
                     class="custom-select" 
                     id="place"
                     name="place" 
                     onChange={this.onPlaceChange}
                     value={this.state.place.toLowerCase()}
                  >
                     {
                        this.props.settings.places.map( place => {
                           return <option value={place.toLowerCase()} key={place}>{place}</option>
                        })
                     }
                  </select>
               </div>
               <div class="input-group mb-2">
                  <div class="input-group-prepend">
                     <label class="input-group-text" for="date">Date</label>
                  </div>
                  <SingleDatePicker 
                     date={this.state.date}
                     onDateChange={this.onDateChange}
                     focused={this.state.focused}
                     onFocusChange={this.onFocusChange}
                     numberOfMonths={1}
                     isOutsideRange={() => false}
                  />
               </div>
               <div class="form-group">
                  <textarea 
                     class="form-control"
                     value={this.state.note}
                     name="note" 
                     placeholder="Add your note"
                     onChange={this.onNoteChange}
                  ></textarea>
               </div>
               <div class="form-group text-right">
                  <Link class="text-danger mr-2" to={'/'}>Cancel</Link>
                  <button class="btn btn-primary">Submit</button>
               </div>
            </form>
         </div>
      )
   }
}

const mapStateToProps = ({ settings }) => ({
   settings: settings
})

export default connect(mapStateToProps)(AddEventForm)