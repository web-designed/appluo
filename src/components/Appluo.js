import React from 'react'
import EventsList from './EventsList'

class Appluo extends React.Component {
   render() {
      return (
         <div>
            <EventsList />
         </div>
      )
   }
}

export default Appluo