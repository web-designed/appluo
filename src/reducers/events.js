//*******************************************************
// EVENTS REDUCERS
//*******************************************************

   const eventsReducerDefaultState = []
   const eventsReducer = (state = eventsReducerDefaultState, action) => {
      switch(action.type){
         case 'ADD_EVENT':
            return state.concat(action.event)
         case 'REMOVE_EVENT':
            return state.filter(({ id }) => id !== action.id)
         case 'EDIT_EVENT':
            return state.map((event) => {
               if(event.id === action.id) {
                  return {
                     ...event,
                     ...action.update
                  }
               } else {
                  return event
               }
            })
         case 'ADD_COMMENT':
            return state.map((event) => {
               if(event.id === action.eventId){
                  return {
                     ...event,
                     comments: [...event.comments, action.comment]
                  }
               } else {
                  return event
               }
            })
         case 'REMOVE_COMMENT':
            return state.map((event) => {
               if(event.id === action.eventId){
                  return {
                     ...event,
                     comments: event.comments.filter((comment) => {
                        return comment.id !== action.commentId
                     })
                  }
               } else {
                  return event
               }
            })
         case 'EDIT_COMMENT':
            return state.map((event) => {
               if(event.id === action.eventId){
                  return {
                     ...event,
                     comments: event.comments.map((comment) => {
                        if(comment.id === action.commentId){
                           return {
                              ...comment,
                              ...action.update
                           }
                        } else {
                           return comment
                        }
                     }) 
                  }
               } else {
                  return event
               }
            })
         default:
            return state
      }
   }

export default eventsReducer