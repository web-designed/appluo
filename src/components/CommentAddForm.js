import React from 'react'
import uuid from 'uuid'
import { connect } from 'react-redux'
import { startAddComment } from '../actions/comments'
import moment from 'moment'

export class CommentAddForm extends React.Component {

   state = {
      error: '',
      commentBody: '',
      commenter: 'kylu'
   }

   onCommentBodyChange = (e) => {
      const commentBody = e.target.value
      this.setState(() => ({ commentBody }))
   }

   handleSubmitComment = (e) => {

      e.preventDefault()

      if(!this.state.commentBody){
         this.setState(() => ({
            error: this.props.notifications.validate
         }))
      } else {
         const newComment = {
            commentBody: this.state.commentBody,
            commenter: this.state.commenter
         }
         //dispatch the action
         this.props.startAddComment(this.props.currentEventId, newComment)

         //clear the comment and the validation error
         this.setState(() => ({
            commentBody: '',
            error: ''
         }))
      }
   }

   render(){
      return (
         <div class="pt-3">
            <form onSubmit={this.handleSubmitComment}>
               <div class="form-group">
                  <textarea 
                     class="form-control bg-light"
                     name="commentBody"
                     value={this.state.commentBody}
                     onChange={this.onCommentBodyChange}
                     placeholder="Add your comment"
                  ></textarea>
                  {this.state.error && <p class="text-danger"><small>{this.state.error}</small></p>}
               </div>
               <div class="form-group">
                  <button class="btn btn-secondary">Add Comment</button>
               </div>
            </form>
         </div>
      )
   }
}

const mapDispatchToProps = dispatch => ({
   startAddComment: (id, newComment) => dispatch(startAddComment(id, newComment))
})

const mapStateToProps = state => ({
   notifications: state.settings.notifications.comment
})

export default connect(mapStateToProps, mapDispatchToProps)(CommentAddForm)