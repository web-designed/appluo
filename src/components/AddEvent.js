import React from 'react'
import AddEventForm from './AddEventForm'

const AddEvent = () => {
   return(
      <div>
         <h1>Add a cleaning event</h1>
         <AddEventForm />
      </div>
   )
}

export default AddEvent