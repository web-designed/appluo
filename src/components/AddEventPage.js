import React from 'react'
import AddEventForm from './AddEventForm'
import { addEvent } from '../actions/events'
import { connect } from 'react-redux'

const AddEventPage = (props) => (
   <div>
      <h1>Add a cleaning</h1>
      <AddEventForm 
         handleSubmit={ (event) => {
            props.onSubmit(event)
            props.history.push('/')
         }} 
         history={props.history} 
      />
   </div>
)

const mapDispatchToProps = (dispatch) => ({
   onSubmit: (expense) => dispatch(addEvent(expense))
})

export default connect(undefined, mapDispatchToProps)(AddEventPage)