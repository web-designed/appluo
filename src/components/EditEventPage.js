import React from 'react'
import AddEventForm from './AddEventForm'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import { editEvent } from '../actions/events'

const EditEventPage = ( props ) => {

   const currentEvent = props.events.find((event) => {
      return event.id === props.match.params.id
   })

   if(currentEvent){
      return(
         <div>
            <h1>Edit your cleaning</h1>
            <AddEventForm 
               history={props.history}  
               cleaner={currentEvent.cleaner}
               place={currentEvent.place}
               cleanedAt={currentEvent.cleanedAt}
               note={currentEvent.note}
               handleSubmit={ (event) => {
                  props.dispatch(editEvent(currentEvent.id, event))
                  props.history.push('/')
               }}
            />
         </div>
      )
   } else {
      return (
         <div>
            <h1>This cleaning event was removed</h1>
            <Link to="/"><button>Dashboard</button></Link>
         </div>
      )
   }

}

const mapStateToProps = (state) => ({
   events: state.events
})

export default connect(mapStateToProps)(EditEventPage)