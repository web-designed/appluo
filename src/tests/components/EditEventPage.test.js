import React from 'react'
import { EditEventPage } from '../../components/EditEventPage'
import { shallow } from 'enzyme'
import events from '../fixtures/events'

let wrapper, history, editEvent

beforeEach(() => {
   editEvent = jest.fn()
   history = { push: jest.fn() }
   wrapper = shallow(<EditEventPage event={events[0]} history={history} editEvent={editEvent} />)
})

test('should render EditEventPage correctly if event doesnt exist', () => {
   const wrapperNoEvent = shallow(<EditEventPage />)
   expect(wrapperNoEvent).toMatchSnapshot()
})

test('should render EditEventPage correctly if event exist', () => {
   expect(wrapper).toMatchSnapshot()
})

test('should handle the handleSubmit function', () => {
   wrapper.find('Connect(AddEventForm)').prop('handleSubmit')(events[0])
   expect(history.push).toHaveBeenLastCalledWith('/')
   expect(editEvent).toHaveBeenLastCalledWith(events[0].id, events[0])
})