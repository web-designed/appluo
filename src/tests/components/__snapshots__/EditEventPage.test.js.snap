// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditEventPage correctly if event doesnt exist 1`] = `
<div
  class="container-fluid"
>
  <div
    class="pt-3 row justify-content-center"
  >
    <div
      class="col-md-6"
    >
      <h3>
        This Cleaning doesn't exist or has been removed
      </h3>
      <Link
        class="btn btn-secondary mr-2"
        replace={false}
        to="/"
      >
        Dashboard
      </Link>
      <Link
        class="btn btn-primary"
        replace={false}
        to="/add"
      >
        Add Cleaning
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render EditEventPage correctly if event exist 1`] = `
<div
  class="container-fluid"
>
  <div
    class="pt-3 row justify-content-center"
  >
    <div
      class="col-md-6"
    >
      <div
        class="card"
      >
        <div
          class="card-header"
        >
          <h4
            class="mb-0"
          >
            Edit your cleaning
          </h4>
        </div>
        <div
          class="card-body"
        >
          <Connect(AddEventForm)
            cleanedAt={0}
            cleaner="Kylu"
            handleSubmit={[Function]}
            history={
              Object {
                "push": [MockFunction],
              }
            }
            note="Alles sauber gemacht"
            place="Bad"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
